kernel/kernel-api/src/main/java/io/delta/kernel/internal/ScanImpl.java

```diff
@@ -309,4 +346,79 @@ private CloseableIterator<FilteredColumnarBatch> applyDataSkipping(
               filteredScanFileBatch.getData(), Optional.of(newSelectionVector));
         });
   }
+
+  /**
+   * Wraps a scan file iterator such that we emit {@link ScanReport} to the engine upon success and
+   * failure. Since most of our scan building code is lazily executed (since it occurs as
+   * maps/filters over an iterator) potential errors don't occur just within `getScanFile`s
+   * execution, but rather may occur as the returned iterator is consumed. Similarly, we cannot
```

Comment:
so the fact that exceptions may come from within `getScanFile` and from external (i.e. as it is consumed) is why we are wrapping the iteration, catching exceptions, reporting them to reporters, and _then_ throwing them, right?

I wonder if you could make this a bit more clear? about the _why_ we are doing this, as opposed to _what_ this method is doing?

Reply:
Not exactly. If an exception originates externally (i.e. within connector code) there is nothing we can do/report. We wrap these methods since most of the work in `getScanFiles` is executed lazily (i.e. when you call `hasNext` or `next`). Since these methods are where the bulk of the execution lies this is also where exceptions may originate (for example, IO errors occur when reading files, which we do lazily in log replay as part of the `hasNext` check). When exceptions occur then, lazily, within these methods, `getScanFiles` has already executed.

Reply:
I think this addresses the why (at least for the error case) but let me know if there's a way to make it more clear

>  Since most of our scan building code is lazily executed (since it occurs as maps/filters over an iterator) potential errors don't occur just within `getScanFile`s execution, but rather may occur as the returned iterator is consumed.

Reply:
I still don't think I get it :) could we have a 5 min sync for you to explain?

Reply:
Updated the docs based on our discussion